src/main/java/io/supertokens/cronjobs/bulkimport/ProcessBulkImportUsers.java

```diff
@@ -0,0 +1,597 @@
+        for (SQLStorage storage : userPoolToStorageMap.values()) {
+            storage.closeConnectionForBulkImportProxyStorage();
+            storage.close();
+        }
+        userPoolToStorageMap.clear();
+    }
+
+    private void processUser(AppIdentifier appIdentifier, BulkImportUser user, BulkImportUserUtils bulkImportUserUtils,
+            BulkImportSQLStorage baseTenantStorage)
+            throws TenantOrAppNotFoundException, StorageQueryException, InvalidConfigException, IOException,
+            DbInitException {
+
+        try {
+            if (Main.isTesting && Main.isTesting_skipBulkImportUserValidationInCronJob) {
+                // Skip validation when the flag is enabled during testing
+            } else {
+                // Validate the user
+                bulkImportUserUtils.createBulkImportUserFromJSON(main, appIdentifier, user.toJsonObject(), user.id);
+            }
+
+            // Since all the tenants of a user must share the storage, we will just use the
+            // storage of the first tenantId of the first loginMethod
+
+            TenantIdentifier firstTenantIdentifier = new TenantIdentifier(appIdentifier.getConnectionUriDomain(),
+                    appIdentifier.getAppId(), user.loginMethods.get(0).tenantIds.get(0));
+
+            SQLStorage bulkImportProxyStorage = (SQLStorage) getProxyStorage(firstTenantIdentifier);
+
+            LoginMethod primaryLM = getPrimaryLoginMethod(user);
+
+            AuthRecipeSQLStorage authRecipeSQLStorage = (AuthRecipeSQLStorage) getProxyStorage(firstTenantIdentifier);
+            // If primaryUserId is not null, it means we may have already processed this user but failed to delete the entry
+            // If the primaryUserId exists in the database, we'll delete the corresponding entry from the bulkImportUser table and proceed to skip this user.
+            if (user.primaryUserId != null) {
+                AuthRecipeUserInfo processedUser = authRecipeSQLStorage.getPrimaryUserById(appIdentifier,
+                        user.primaryUserId);
+
+                if (processedUser != null && isProcessedUserFromSameBulkImportUserEntry(processedUser, user)) {
+                    baseTenantStorage.startTransaction(con2 -> {
+                        baseTenantStorage.deleteBulkImportUser_Transaction(appIdentifier, con2, user.id);
+                        return null;
+                    });
+                    return;
+                }
+            }
+
+            bulkImportProxyStorage.startTransaction(con -> {
+                try {
+                    for (LoginMethod lm : user.loginMethods) {
+                        processUserLoginMethod(appIdentifier, bulkImportProxyStorage, lm);
+                    }
+
+                    createPrimaryUserAndLinkAccounts(main, appIdentifier, bulkImportProxyStorage, user, primaryLM);
+                    createUserIdMapping(main, appIdentifier, user, primaryLM);
```

Comment:
creation of external user id should happen before calling createPrimaryUserAndLinkAccounts? Cause the `AuthRecipe.createPrimaryUser` takes in the external user id as an input.

Reply:
`AuthRecipe.createPrimaryUser` doesn't work with externalUserId.